### PR TITLE
Move update templates methods to core

### PR DESCRIPTION
**Pull request by jdiazgon**
Needed for #819 as we need to make available the methods related to the download of templates jar.

Changes:

* Moved all the needed methods for updating the jars to the core.
* Fixed some bugs, the updating of templates was not done correctly.
* Implemented a method to check if our current jar is outdated. May be useful in future with the CLI, to ask our user whether we should update the latest templates.

@devonfw/cobigen

### CODE REVIEW

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+import java.nio.file.StandardCopyOption;
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.devonfw.cobigen.api.constants.TemplatesJarConstants;
+
+/**
+ *
+ */
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: missing javadoc

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            String lowercaseName = name.toLowerCase();
+            String regex = TemplatesJarConstants.JAR_FILE_REGEX_NAME;
+
+            Pattern p = Pattern.compile(regex);
+            Matcher m = p.matcher(lowercaseName);
+            if (m.find()) {
+                return true;
+            } else {
+                return false;
+            }
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: just return m.find()

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            String lowercaseName = name.toLowerCase();
+            String regex = TemplatesJarConstants.SOURCES_FILE_REGEX_NAME;
+
+            Pattern p = Pattern.compile(regex);
+            Matcher m = p.matcher(lowercaseName);
+            if (m.find()) {
+                return true;
+            } else {
+                return false;
+            }
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: return m.find()

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+     *            directory where the templates jar are located
+     * @return fileName Name of the file downloaded
+     * @throws MalformedURLException
+     *             {@link MalformedURLException} occurred
+     * @throws IOException
+     *             {@link IOException} occurred
+     */
+    public static String downloadJar(boolean isDownloadSource, File templatesDirectory)
+        throws MalformedURLException, IOException {
+        String mavenUrl =
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: maybe we should at least make this groupId and artifactId and version also configurable, meaning, we have a second function providing the ability to download a different jar and renaming this function to downloadLatestDevon4jTemplates.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        Matcher m = matchJarVersion(fileName, isDownloadSource);
+        if (m.find() == false || m.group(2).isEmpty()) {
+            // Maybe the jar is corrupted, let's update it
+            return true;
+        } else {
+            // Split the version number because it contains dots e.g. 3.1.0
+            int[] versionNumbers = Arrays.stream(m.group(2).split("\\.")).mapToInt(Integer::parseInt).toArray();
+
+            // We do the same for the latest jar in Maven, therefore we need to download it
+            HttpURLConnection conn = initializeConnection(mavenUrl);
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: could you please check, that HttpURLConnect might implement Closeable and thus should be in try( ... ) {} block to be automatically closed.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+     *             if the URL is invalid
+     * @throws IOException
+     *             if we could not connect properly
+     * @throws ProtocolException
+     *             if the request protocol is invalid
+     */
+    private static HttpURLConnection initializeConnection(String mavenUrl)
+        throws MalformedURLException, IOException, ProtocolException {
+        URL url = new URL(mavenUrl);
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: same here. please check for closeable

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java (hunk 5 above)
Comment (maybeec) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: ok,  just googled it. You are right, but we should do it with `conn.getInputStream()` definitively.

Comment (jdiazgon) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: Okay, done!

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+     *            true if downloading source jar file
+     * @param templatesDirectory
+     *            directory where the templates jar are located
+     * @return fileName Name of the file downloaded
+     * @throws MalformedURLException
+     *             {@link MalformedURLException} occurred
+     * @throws IOException
+     *             {@link IOException} occurred
+     */
+    public static String downloadLatestDevon4jTemplates(boolean isDownloadSource, File templatesDirectory)
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: why not reusing download Jar here?

Comment (jdiazgon) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java]: Wops, you are right

**Pull request merged**